### PR TITLE
e2e: add test for cpnp replication

### PR DESCRIPTION
**Pull request by GiedriusS**
Add e2e test that shows that cpnp replication works. Remove some unused code.

### CODE REVIEW

File: test/e2e/receive_test.go [hunk 1]
```diff
@@ -1152,3 +1152,42 @@ func TestReceiveExtractsTenant(t *testing.T) {
...
+					},
+					Samples: []prompb.Sample{
+						{Value: 1, Timestamp: time.Now().UnixMilli()},
+					},
+				},
+			},
+		})
+	}))
+
+	testutil.Ok(t, i.WaitSumMetricsWithOptions(e2emon.Equals(0), []string{"prometheus_tsdb_blocks_loaded"}, e2emon.WithLabelMatchers(matchers.MustNewMatcher(matchers.MatchEqual, "tenant", "default-tenant")), e2emon.WaitMissingMetrics()))
```
Comment (pedro-stanaka) [inline test/e2e/receive_test.go:1195]: can we try querying the metric directly?

**Review by pedro-stanaka — APPROVED** [review]

File: test/e2e/receive_test.go (hunk 1 above)
Comment (GiedriusS) [reply test/e2e/receive_test.go:1195]: Added this

**Comment by GiedriusS** [comment]
Very weird - writes succeed and I can see the timestamps/values coming in as expected, and the series is in the WAL but I can't query them :/ something is wrong with the writer but I cannot spot anything wrong so far.

**Comment by fpetkovski** [comment]
I added some logging to the writer and I can also see 
```
Appending sample {__name__="myself"} 1729090293939 1
err: <nil>
```

Maybe something with the timestamp format?

**Comment by fpetkovski** [comment]
Also seeing metrics in the ingestor 

```
# HELP prometheus_tsdb_head_samples_appended_total Total number of appended samples.
# TYPE prometheus_tsdb_head_samples_appended_total counter
prometheus_tsdb_head_samples_appended_total{tenant="default-tenant",type="float"} 1
prometheus_tsdb_head_samples_appended_total{tenant="default-tenant",type="histogram"} 0
```

**Comment by fpetkovski** [comment]
@GiedriusS changing [this line](https://github.com/thanos-io/thanos/pull/7659/files#diff-9da5aaa9d6622a1eae7a0a29cb9638d491e43e0724edb2b2eb7f6c2f7af3b0edR68) from  `unsafe.String(&b[0], len(b))` to `string(b)` makes the test pass. I assume that `labels.Copy` [here](https://github.com/thanos-io/thanos/pull/7659/files#diff-c23d581dac9bba662216511c48c42a72b727ae8575284d2373435663e4d4e749R84) only does a shallow copy of labels and actual name and value bytes get lost at some point.

For now we can either remove the unsafe conversion, or we can have a deep copy of labels which re-allocates the name and value strings.

**Comment by GiedriusS** [comment]
Added a `strings.Clone()` where it is really necessary. I think it is preferable to always making a copy because creating a new series is a much rarer occurrence than adding new samples to an existing series.

File: pkg/receive/capnproto_writer.go [hunk 2]
```diff
@@ -72,6 +73,13 @@ func (r *CapNProtoWriter) Write(ctx context.Context, tenantID string, wreq *writ
 		ref, lset = getRef.GetRef(series.Labels, series.Labels.Hash())
 		if ref == 0 {
 			lset = series.Labels.Copy()
+			// NOTE(GiedriusS): do a deep copy because the labels are reused in the capnp message.
+			// Creation of new series is much rarer compared to adding extra samples
+			// to an existing series.
+			for i := range lset {
```
Comment (fpetkovski) [inline pkg/receive/capnproto_writer.go:79]: We need to use `lset.Range`

Comment (fpetkovski) [reply pkg/receive/capnproto_writer.go:79]: I'll fix this on my branch.

**Pull request merged**